### PR TITLE
use a priority ordered list of addresses in autonat v2

### LINKED ISSUE #539: Proposal: Autonat V2 accept a list of addresses in priority order
In discussion with @marten-seemann he suggested: 

> Maybe the client could send a list of addresses it wants dialed, in descending priority. The receiver would pick exactly one of them and dial it

the proposal is:

Client has a list of addresses it wants to determine reachability for. In a dialRequest they send this list of address in priority order to the autonat server. The autonat server dials the first address it is capable of dialing and reports success or failure for that address. 

This will be helpful in cases where we are interested in determining reachability for a newly address transport for which support in the network is low. Many requests for this new transport address will be rejected because the server cannot dial this transport. This proposal allows us to reuse those requests to check reachability for other addresses. 

For some history this suggestion is an improvement over the scheme suggested here https://github.com/libp2p/specs/pull/538#discussion_r1162730358 . This scheme aimed to solve the same problem.

### PR DESCRIPTION
**Pull request by sukunrt**
Closes #539

### CODE REVIEW

File: autonat/autonat-v2.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 | Lifecycle Stage | Maturity                 | Status | Latest Revision |
 |-----------------|--------------------------|--------|-----------------|
-| 1A              | Working Draft            | Active | r1, 2023-04-12  |
+| 1A              | Working Draft            | Active | r2, 2023-04-15  |
```
Comment (marten-seemann) [inline autonat/autonat-v2.md:6]: No need to bump the revision for unmerged specs.

File: autonat/autonat-v2.md [hunk 2]
```diff
@@ -19,71 +19,89 @@ Interest Group: [@marten-seemann], [@marcopolo], [@mxinden]
...
-Using `autonat v2` nodes can build an address pipeline where they can test 
-individual addresses discovered by different sources like identify, upnp 
-mappings, circuit addresses etc for reachability.
+In `autonat v2` client sends a priority ordered list of addresses. On receiving
+this list the server dials the first address on the list that it is capable of
+dialing. `autonat v2` allows nodes to determine reachability for individual
+addresses. Using `autonat v2` nodes can build an address pipeline where they can
+test individual addresses discovered by different sources like identify, upnp
+mappings, circuit addresses etc for reachability. Having a priority ordered list
+of addresses provides the ability to verify low priority addresses.
```
Comment (marten-seemann) [inline autonat/autonat-v2.md:34]: Should this mention that this helps with the rollout of new transports that are only supported by few nodes in the network?

File: autonat/autonat-v2.md [hunk 3]
```diff
@@ -19,71 +19,89 @@ Interest Group: [@marten-seemann], [@marcopolo], [@mxinden]
...
+this list contains an address and a fixed64 nonce. The list is ordered in
+descending order of priority for verfication.
 
-Upon completion of the dial attempt, the peer sends a `DialResponse` to the 
-initiator node on the `/libp2p/autonat/2.0.0/dial-request` stream.
+Upon receiving this message the peer attempts to dial the first address from the
+list that it is capable of dialing. It dials this address, opens a stream with
+Protocol ID `/libp2p/autonat/2.0.0/dial-attempt` and sends a `DialAttempt`
+message with the nonce received in the corresponding `AddressDialRequest`. The
+peer MUST dial the first address in the list it is capable of dialing. The peer
```
Comment (marten-seemann) [inline autonat/autonat-v2.md:59]: The logic is correct, but we shouldn't use a MUST here. The peer can still refuse to dial, e.g. due to rate-limiting.

File: autonat/autonat-v2.md [hunk 4]
```diff
@@ -127,37 +151,40 @@ message Message {
...
+      E_DIAL_ERROR              = 100;
+      E_DIAL_REFUSED            = 101;
+      E_TRANSPORT_NOT_SUPPORTED = 102;
+      E_BAD_REQUEST             = 200;
+      E_INTERNAL_ERROR          = 300;
+    }
+
+    ResponseStatus status = 1;
+    string statusText = 2;
+    bytes addr = 3;
```
Comment (marten-seemann) [inline autonat/autonat-v2.md:185]: Is there any advantage to sending the entire address? This could be the list index of the address in the `DialRequest`.

**Review by marten-seemann — COMMENTED** [review]
I like it!

File: autonat/autonat-v2.md (hunk 1 above)
Comment (sukunrt) [reply autonat/autonat-v2.md:6]: ah thanks!

File: autonat/autonat-v2.md [hunk 5]
```diff
@@ -19,71 +19,89 @@ Interest Group: [@marten-seemann], [@marcopolo], [@mxinden]
...
+peer MUST dial the first address in the list it is capable of dialing. The peer
+MUST NOT dial any address not in the list of addresses sent in the request. The
+peer MUST ignore errors on `/libp2p/autonat/2.0.0/dial-attempt` stream
+
+Upon completion of the dial attempt, the peer sends a `DialResponse` message to
+the initiator node on the `/libp2p/autonat/2.0.0/dial-request` stream with the
+address that it attempted to dial and the appropriate `ResponseStatus`. see
+[Requirements For ResponseStatus](#requirements-for-responsestatus)
+
+The initiator SHOULD check that the nonce received in the `DialAttempt` is the
```
[the remark this answers is not in the document]
Comment (sukunrt) [reply autonat/autonat-v2.md:68]: https://github.com/libp2p/specs/pull/538#discussion_r1167452081

I'll change this to MUST.

File: autonat/autonat-v2.md [hunk 6]
```diff
@@ -19,71 +19,89 @@ Interest Group: [@marten-seemann], [@marcopolo], [@mxinden]
...
 
-Upon completion of the dial attempt, the peer sends a `DialResponse` to the 
-initiator node on the `/libp2p/autonat/2.0.0/dial-request` stream.
+Upon receiving this message the peer attempts to dial the first address from the
+list that it is capable of dialing. It dials this address, opens a stream with
+Protocol ID `/libp2p/autonat/2.0.0/dial-attempt` and sends a `DialAttempt`
+message with the nonce received in the corresponding `AddressDialRequest`. The
+peer MUST dial the first address in the list it is capable of dialing. The peer
+MUST NOT dial any address not in the list of addresses sent in the request. The
+peer MUST ignore errors on `/libp2p/autonat/2.0.0/dial-attempt` stream
```
[the remark this answers is not in the document]
Comment (sukunrt) [reply autonat/autonat-v2.md:61]: here https://github.com/libp2p/specs/pull/541#discussion_r1170995752

File: autonat/autonat-v2.md (hunk 2 above)
Comment (sukunrt) [reply autonat/autonat-v2.md:34]: I added text that explains this case.

File: autonat/autonat-v2.md (hunk 3 above)
Comment (sukunrt) [reply autonat/autonat-v2.md]: fixed. I removed MUST and added this requirement to MUST NOT phrase.

File: autonat/autonat-v2.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (sukunrt) [reply autonat/autonat-v2.md]: fixed

File: autonat/autonat-v2.md (hunk 4 above)
Comment (sukunrt) [reply autonat/autonat-v2.md]: changing this to index, I think that would also be easier for implementations

File: autonat/autonat-v2.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (sukunrt) [reply autonat/autonat-v2.md]: fixed.

**Pull request merged**